### PR TITLE
Add `prefect api` command for direct API access

### LINKED ISSUE #19130: Feature: Add `prefect api` CLI command for direct API access
# Feature: `prefect api` CLI command

## Problem

The Prefect CLI covers a lot of common operations, but there are times when you need to make an API call that doesn't have a dedicated command yet. Right now, your options are to use `curl` or write a Python script, which means manually managing authentication, constructing URLs, and handling the response format.

## Proposed Solution

Add a `prefect api` command that works like `gh api` for GitHub - a generic HTTP client for the Prefect API that uses your current configuration (API key, API URL) and handles the common patterns for you.

### Basic requests

```bash
# List flows (many endpoints use POST for filtering)
prefect api POST /flows/filter

# Get a specific flow
prefect api GET /flows/550e8400-e29b-41d4-a716-446655440000
```

The HTTP method is a required positional argument because many Prefect API endpoints use POST for listing/filtering operations. The command automatically uses your configured API key and URL.

### Smart path handling

The command would automatically prepend the appropriate prefix based on your configuration:

**For Prefect Cloud users:**
- Automatically adds `/api/accounts/<account_id>/workspaces/<workspace_id>` prefix
- So `prefect api POST /flows/filter` becomes `POST /api/accounts/abc-123/workspaces/xyz-789/flows/filter`

**For OSS users:**
- Automatically adds `/api` prefix
- So `prefect api POST /flows/filter` becomes `POST /api/flows/filter`

**For account-level operations (Cloud only):**
```bash
# Use a flag to skip workspace prefix for account-level operations
prefect api GET /workspaces --account
```

**For API root-level operations:**
```bash
# Some endpoints sit at the API root (e.g., /api/me)
# Use a flag to access these directly
prefect api GET /me --root
prefect api GET /health --root
```

### Query parameters

```bash
# Simple query parameter
prefect api GET '/flows/some-id?include=settings'

# Multiple query parameters (note the quotes to handle &)
prefect api GET '/work_pools?limit=10&offset=20'

# Works with any method
prefect api POST '/flows/filter?limit=100' --data '{"tags": ["production"]}'
```

### POST/PUT/PATCH/DELETE requests

```bash
# Create a deployment
prefect api POST /deployments --data '{"name": "my-deployment", "flow_id": "..."}'

# Update via PATCH
prefect api PATCH /deployments/abc-123 --data '{"is_schedule_active": false}'

# Delete
prefect api DELETE /flows/abc-123
```

### Alternative syntax (curl/gh compatibility)

```bash
# Can use -X if you prefer (only when method not specified positionally)
prefect api /deployments -X POST --data '{"name": "my-deployment"}'

# Custom headers with -H/--header
prefect api POST /flows/filter -H "X-Custom-Header: value" --data '{}'

# Multiple headers
prefect api POST /deployments \
  -H "X-Request-ID: 12345" \
  -H "X-Custom: value" \
  --data '{"name": "test"}'
```

### Input from stdin

```bash
# Pipe JSON data into the command (automatically detected)
cat deployment.json | prefect api POST /deployments

# Or from heredoc
prefect api POST /deployments <<EOF
{
  "name": "my-deployment",
  "flow_id": "550e8400-e29b-41d4-a716-446655440000"
}
EOF

# Explicit --data flag still supported for inline or file input
prefect api POST /deployments --data '{"name": "my-deployment", "flow_id": "..."}'
prefect api POST /deployments --data @deployment.json
```

### Pretty-printed responses

By default, format JSON responses for readability:

```bash
$ prefect api POST /flows/filter --data '{}'
[
  {
    "id": "550e8400-e29b-41d4-a716-446655440000",
    "name": "my-flow",
    "created": "2024-01-15T10:30:00Z"
  }
]
```

### Verbose output for debugging

```bash
# Show request/response headers and other debug info
prefect api POST /flows/filter --data '{}' --verbose

# Output would include:
# > POST /api/accounts/.../workspaces/.../flows/filter
# > Authorization: Bearer ...
# > Accept: application/json
# > Content-Type: application/json
# < HTTP/1.1 200 OK
# < Content-Type: application/json
# < ...
# [response body]
```

### Error handling

Show helpful error messages when things go wrong:

```bash
$ prefect api GET /flows/invalid-id
Error: 404 Not Found
{
  "detail": "Flow not found"
}

$ prefect api POST /flows/filter --data '{}' (with invalid API key)
Error: 401 Unauthorized
Check your API key configuration with: prefect config view

$ prefect api GET /flows/invalid-id --verbose
> GET /api/accounts/.../workspaces/.../flows/invalid-id
> Authorization: Bearer ...
> Accept: application/json
< HTTP/1.1 404 Not Found
< Content-Type: application/json
< Content-Length: 35
Error: 404 Not Found
{
  "detail": "Flow not found"
}
```

### Exit codes

The command should exit with meaningful status codes for scripting:

```bash
# Success (2xx responses)
$ prefect api GET /flows/some-id
# exits with 0

# Client errors (4xx responses)
$ prefect api GET /flows/invalid-id
# exits with 4

# Server errors (5xx responses)
$ prefect api GET /flows/some-id (when server is having issues)
# exits with 5

# Network/connection errors
$ prefect api GET /flows/some-id (when API is unreachable)
# exits with 7

# Authentication errors (401, 403)
$ prefect api GET /flows/some-id (with invalid key)
# exits with 3
```

This allows for reliable scripting:
```bash
if prefect api GET /flows/$FLOW_ID > /dev/null 2>&1; then
  echo "Flow exists"
else
  echo "Flow not found or error occurred: $?"
fi
```

## Example Use Cases

1. **Testing API endpoints during development**
   ```bash
   prefect api GET /work_pools/my-pool/workers --verbose
   ```

2. **Quick data exports**
   ```bash
   prefect api POST /flows/filter --data '{}' | jq '.[] | {id, name}' > flows.json
   ```

3. **Scripting operations not yet in the CLI**
   ```bash
   # Bulk operations
   for flow_id in $(cat flow_ids.txt); do
     prefect api DELETE /flows/$flow_id
   done
   ```

4. **Account-level operations on Prefect Cloud**
   ```bash
   prefect api GET /workspaces --account
   prefect api GET /service_accounts --account
   ```

## Questions for Feedback

1. What should the flag names be? I'm thinking:
   - `--data` for POST body
   - `--account` for account-level operations (Cloud only)
   - `--root` for API root-level operations (e.g., `/api/me`)
   - `-v/--verbose` for debug output
2. Should we support other output formats (e.g., `--raw` for unpretty JSON, `--format yaml`)?
3. Should there be a `--dry-run` mode that shows what request would be made without executing it?
4. Any other features from `gh api` or similar tools that would be useful?

## Design Philosophy

This tool is intentionally minimal - it's direct API access with just enough convenience (auth, URL construction, pretty output) to be useful. Features like automatic pagination, response filtering, or request retries are explicitly out of scope. Users who need those should use the Python SDK or write scripts around this command.

## Related Tools

- GitHub CLI: `gh api`
- AWS CLI: `aws <service> <operation>`
- Stripe CLI: `stripe <resource> <action>`

### PR DESCRIPTION
**Pull request by chrisguidry**
This adds a `prefect api` command similar to `gh api` for GitHub - providing a simple HTTP client for the Prefect API with automatic auth and URL handling.

## What it does

Makes it easy to call API endpoints that don't have dedicated CLI commands yet, without having to manually construct URLs or manage authentication:

```bash
# List flows using the filter endpoint
prefect api POST /flows/filter --data '{"limit": 10}'

# Get a specific resource
prefect api GET /flows/abc-123

# Works with stdin
echo '{"limit": 5}' | prefect api POST /deployments/filter

# Verbose mode for debugging
prefect api GET /flows --verbose
```

## Key features

- HTTP method as positional argument (matches Prefect's POST-heavy filter endpoints)
- Smart path handling - automatically adds workspace/account/api prefixes
- Multiple input methods: `--data`, `@filename`, or auto-detected stdin
- Custom headers with `-H` (curl/gh compatible)
- Pretty-printed JSON (compact when piped to other tools)
- Meaningful exit codes for scripting (0=success, 3=auth, 4=client error, 5=server, 7=network)
- Defaults to `{}` for POST/PUT/PATCH without data (filter endpoints just work)
- Uses PrefectClient under the hood for ephemeral servers and connection pooling

## Implementation notes

- URL construction uses stdlib `urllib.parse` for correctness
- Settings accessed via `get_current_settings()` per Prefect conventions
- Leverages PrefectClient's `_client` for raw requests while keeping all its infrastructure benefits
- 27 tests covering all features and edge cases

Closes #19130

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/prefect/cli/api.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    """
+    if not path.startswith("/"):
+        path = f"/{path}"
+
+    parsed = urlparse(api_url)
+    is_cloud = cloud_api_url and api_url.startswith(cloud_api_url)
+
+    if root:
+        new_path = f"/api{path}"
+        return urlunparse((parsed.scheme, parsed.netloc, new_path, "", "", ""))
```
Comment (desertaxle) [inline src/prefect/cli/api.py]: I think this root case should only be valid if the URL is a Cloud URL, since everything is inherently root in a self-hosted server, and a self-hosted server route may have a different base path than `/api`.

File: src/prefect/cli/api.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        req = response.request
+        output_parts.append(f"> {req.method} {req.url}")
+        for key, value in req.headers.items():
+            if key.lower() == "authorization":
+                value = "Bearer ***"
+            output_parts.append(f"> {key}: {value}")
+        output_parts.append("")
+
+        output_parts.append(
+            f"< HTTP/1.1 {response.status_code} {response.reason_phrase}"
```
Comment (desertaxle) [inline src/prefect/cli/api.py]: Should this change if HTTP2 is enabled in the client?

File: src/prefect/cli/api.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+        $ prefect api GET /flows --verbose
+
+        # Account-level operation (Cloud)
+        $ prefect api GET /workspaces --account
+
+        # API root level
+        $ prefect api GET /me --root
+    """
+    if method_or_path is None:
+        return
```
Comment (desertaxle) [inline src/prefect/cli/api.py]: Should this exit with an error?

File: src/prefect/cli/api.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        Optional[str],
+        typer.Argument(help="HTTP method (GET, POST, etc.) or path if using -X flag"),
+    ] = None,
+    path_arg: Annotated[
+        Optional[str], typer.Argument(help="API path (e.g., /flows, /flows/filter)")
+    ] = None,
+    method: Annotated[
+        Optional[str],
+        typer.Option("-X", "--method", help="HTTP method (alternative to positional)"),
+    ] = None,
```
Comment (desertaxle) [inline src/prefect/cli/api.py]: This option seems to add a fair amount of complexity. Could we have only one way to define the method?

File: src/prefect/cli/api.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+    request_headers = {"Accept": "application/json"}
+    if api_key:
+        request_headers["Authorization"] = f"Bearer {api_key}"
+    if body_data is not None:
+        request_headers["Content-Type"] = "application/json"
+    request_headers.update(custom_headers)
+
+    try:
+        async with get_client() as client:
+            response = await client._client.request(
```
Comment (desertaxle) [inline src/prefect/cli/api.py]: You should be able to do this to avoid accessing a privatish attribute.
```suggestion
            response = await client.request(
```

File: src/prefect/cli/api.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+                method=http_method,
+                url=url,
+                json=body_data if body_data is not None else None,
+                headers=request_headers,
+            )
+            response.raise_for_status()
+
+            output = format_output(response, verbose)
+            if output:
+                print(output)
```
Comment (desertaxle) [inline src/prefect/cli/api.py]: Could we use `rich` for output?

File: src/prefect/cli/api.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+
+import json
+import sys
+from pathlib import Path
+from typing import Optional
+from urllib.parse import urljoin, urlparse, urlunparse
+
+import httpx
+import typer
+from typing_extensions import Annotated
```
Comment (zzstoatzz) [inline src/prefect/cli/api.py]: [`typing.Annotated` was added in 3.9 i think](https://docs.python.org/3/library/typing.html#typing.Annotated)

File: tests/cli/test_api.py [hunk 8]
```diff
@@ -0,0 +1,613 @@
...
+from prefect.settings import (
+    PREFECT_API_URL,
+    PREFECT_CLOUD_API_URL,
+    temporary_settings,
+)
+from prefect.testing.cli import invoke_and_assert
+
+
+@pytest.fixture
+def mock_httpx_client(monkeypatch):
```
Comment (desertaxle) [inline tests/cli/test_api.py]: Can we use `respx` for mocking here?

File: src/prefect/cli/api.py [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+        if ":" not in header:
+            exit_with_error(
+                f"Invalid header format: {header!r}. Use 'Key: Value' format."
+            )
+        key, value = header.split(":", 1)
+        headers[key.strip()] = value.strip()
+    return headers
+
+
+def parse_data(data: str | None) -> dict | str | None:
```
Comment (zzstoatzz) [inline src/prefect/cli/api.py]: can we have complete types for `dict`?

File: src/prefect/cli/api.py [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+        $ prefect api POST /flows/filter -H "X-Custom: value" --data '{}'
+
+        # Verbose output
+        $ prefect api GET /flows --verbose
+
+        # Account-level operation (Cloud)
+        $ prefect api GET /workspaces --account
+
+        # API root level
+        $ prefect api GET /me --root
```
Comment (zzstoatzz) [inline src/prefect/cli/api.py:178]: should these be wrapped in triple backticks?

File: src/prefect/cli/api.py (hunk 1 above)
Comment (chrisguidry) [reply src/prefect/cli/api.py]: Oh great call great call, yes

File: tests/cli/test_api.py (hunk 8 above)
Comment (zzstoatzz) [inline tests/cli/test_api.py]: nit: can we have complete types for fixtures in new test suites?

File: src/prefect/cli/api.py (hunk 2 above)
Comment (chrisguidry) [reply src/prefect/cli/api.py]: HA this appears to be too eagerly emulating `curl`'s output, thanks for flagging that, it's kinda useless to have that, I'll make this better

File: src/prefect/cli/api.py (hunk 3 above)
Comment (chrisguidry) [reply src/prefect/cli/api.py]: Yeah I'm gonna nix the `-X`/`--method` argument, I was waffling on that a lot and this tipped me over

File: src/prefect/cli/api.py (hunk 4 above)
Comment (chrisguidry) [reply src/prefect/cli/api.py]: aye

File: src/prefect/cli/api.py (hunk 1 above)
Comment (chrisguidry) [reply src/prefect/cli/api.py]: I think I want to keep the behavior of `--root` in all cases, but it should just mean "root of the API URL", it shouldn't have the hard-coded `/api` to it

**Comment by chrisguidry** [comment]
Ahh @desertaxle, `client.requset` doesn't let us "escape" up to account-level endpoints for Cloud URLs, it's setup to take workspace/server-scoped relative paths only.  Will have to keep `_client` for now

**Comment by desertaxle** [comment]
> Ahh @desertaxle, `client.requset` doesn't let us "escape" up to account-level endpoints for Cloud URLs, it's setup to take workspace/server-scoped relative paths only. Will have to keep `_client` for now

@chrisguidry could you use a `CloudClient` for those account-level requests?

File: src/prefect/cli/api.py (hunk 1 above)
Comment (chrisguidry) [reply src/prefect/cli/api.py]: > I think I want to keep the behavior of `--root` in all cases, but it should just mean "root of the API URL", it shouldn't have the hard-coded `/api` to it

Nope, no I don't, I'm doing what you said :D

File: src/prefect/cli/api.py (hunk 7 above)
Comment (chrisguidry) [reply src/prefect/cli/api.py]: My training data cutoff is 2017

File: src/prefect/client/cloud.py [hunk 11]
```diff
@@ -186,6 +186,35 @@ def __exit__(self, *_: object) -> NoReturn:
     async def get(self, route: str, **kwargs: Any) -> Any:
         return await self.request("GET", route, **kwargs)
 
+    async def raw_request(
```
Comment (chrisguidry) [inline src/prefect/client/cloud.py:191]: @desertaxle and @zzstoatzz, some future feedback, there's a confusing discrepancy between the `PrefectClient.request` and `CloudClient.request` that we should resolve.  For now I added this new `raw_request` method to `CloudClient`, but in the fullness of time I think we should make both of them use the same `request` method from `BaseClient` and call it a day
